packages/base-driver/lib/basedriver/core.js

```diff
@@ -153,14 +153,26 @@ class DriverCore {
 
   get log () {
     if (!this._log) {
-      const instanceName = `${this.constructor.name}@${node
-        .getObjectId(this)
-        .substring(0, 4)}`;
-      this._log = logger.getLogger(() =>
-        this.sessionId
-          ? `${instanceName} (${this.sessionId.substring(0, 8)})`
-          : instanceName,
-      );
+      const instanceName = `${this.constructor.name}@${node.getObjectId(this).substring(0, 4)}`;
+      // We don't want the self reference to be captured inside the below closure
+      // to avoid possible memory leaks,
+      // but NodeJS started supporting WeakRef only since v. 14.6
+      /**
+       * @type {WeakRef<DriverCore>|DriverCore}
+       */
+      const self = global.WeakRef ? new global.WeakRef(this) : this;
+      this._log = logger.getLogger(() => {
+        let sessionId;
+        if (self instanceof DriverCore) {
+          sessionId = self.sessionId;
+        } else if (global.WeakRef && self instanceof global.WeakRef) {
+          const ref = self.deref();
+          if (ref?.sessionId) {
+            sessionId = ref.sessionId;
+          }
+        }
+        return sessionId ? `${instanceName} (${sessionId.substring(0, 8)})` : instanceName;
+      });
```

Comment:
I don't think `WeakRef` is necessary here.  In fact, we should probably [avoid use](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef#avoid_where_possible) unless there is no other possible way to do so.

What we're trying to do here is hand a prefix to the logger.  Currently, we're passing a function, but it can just be a string.  This prefix may be dynamic, based upon the current `sessionId` (if any).

What if, upon session creation/destruction (in `createSession` and `deleteSession` of `BaseDriverCore`), we did something like this:

```js
// this would also probably need to be in AppiumDriver's createSession(), which is *not the same* function
async createSession(stuff) {
  this.sessionId = getNewSessionId();
  const logPrefix = createLogPrefix() // does the instanceName stuff w/ the ternary operator, etc.

  // we could do this:
  this._log = logger.getLogger(logPrefix);

  // OR refactor the logger to expose a `setPrefix` method:
  this._log.setPrefix(logPrefix);
}
```

We could create a setter for `sessionId` which did the log-prefix dance as well.

This would only create a potential leak if a driver implementation held on to the old value `this._log` for too long.  But a driver could create a potential leak in myriad ways which we cannot prevent.

Reply:
If we do this, we should also add a test that shows the prefix gets updated.

Reply:
changed to add/remove the prefix in create/deleteSession